File: app/src/main/java/tn/esprit/wayfinder/viewmodels/DestinationVideoViewModel.kt
package tn.esprit.wayfinder.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import tn.esprit.wayfinder.models.*
import tn.esprit.wayfinder.presentation.destinationvideo.DestinationVideoRepository

sealed class DestinationVideoUiState {
    object Idle : DestinationVideoUiState()
    object Loading : DestinationVideoUiState()
    data class Success(val destinations: List<DestinationWithVideoStatus>) : DestinationVideoUiState()
    data class Error(val message: String) : DestinationVideoUiState()
}

sealed class VideoGenerationState {
    object Idle : VideoGenerationState()
    object Generating : VideoGenerationState()
    data class Success(val message: String) : VideoGenerationState()
    data class Error(val message: String) : VideoGenerationState()
}

class DestinationVideoViewModel(
    private val destinationVideoRepository: DestinationVideoRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow<DestinationVideoUiState>(DestinationVideoUiState.Idle)
    val uiState: StateFlow<DestinationVideoUiState> = _uiState.asStateFlow()

    private val _generationState = MutableStateFlow<VideoGenerationState>(VideoGenerationState.Idle)
    val generationState: StateFlow<VideoGenerationState> = _generationState.asStateFlow()

    fun loadUserDestinations(userId: String) {
        viewModelScope.launch {
            try {
                _uiState.value = DestinationVideoUiState.Loading
                val response = destinationVideoRepository.getUserDestinations(userId)
                _uiState.value = DestinationVideoUiState.Success(response.destinations)
            } catch (e: Exception) {
                _uiState.value = DestinationVideoUiState.Error(parseError(e))
            }
        }
    }

    fun generateVideo(userId: String, destination: String) {
        viewModelScope.launch {
            try {
                _generationState.value = VideoGenerationState.Generating
                val response = destinationVideoRepository.generateDestinationVideo(userId, destination)
                _generationState.value = VideoGenerationState.Success(response.message)
                // Reload destinations to update status
                loadUserDestinations(userId)
            } catch (e: Exception) {
                _generationState.value = VideoGenerationState.Error(parseError(e))
            }
        }
    }

    fun checkVideoStatus(userId: String, destination: String) {
        viewModelScope.launch {
            try {
                val status = destinationVideoRepository.getDestinationVideoStatus(userId, destination)
                // Update the destination status in the list if needed
                val currentState = _uiState.value
                if (currentState is DestinationVideoUiState.Success) {
                    val updatedDestinations = currentState.destinations.map { dest ->
                        if (dest.destination == destination) {
                            dest.copy(
                                videoStatus = status.status,
                                videoUrl = status.videoUrl,
                                imageCount = status.imageCount
                            )
                        } else {
                            dest
                        }
                    }
                    _uiState.value = DestinationVideoUiState.Success(updatedDestinations)
                }
            } catch (e: Exception) {
                android.util.Log.e("DestinationVideoViewModel", "Error checking video status: ${e.message}")
            }
        }
    }

    private fun parseError(throwable: Throwable): String {
        return when (throwable) {
            is retrofit2.HttpException -> {
                val errorBody = throwable.response()?.errorBody()?.string()
                errorBody ?: "HTTP ${throwable.code()}: ${throwable.message()}"
            }
            is java.net.UnknownHostException -> "Impossible de se connecter au serveur. Vérifiez votre connexion Internet."
            is java.net.SocketTimeoutException -> "Le serveur met trop de temps à répondre. Veuillez réessayer."
            is java.net.ConnectException -> "Impossible de se connecter au serveur. Vérifiez votre connexion Internet."
            else -> throwable.message ?: "Une erreur inattendue est survenue"
        }
    }
}

File: WayFinder/ViewModels/DestinationVideoViewModel.swift
import Foundation
import SwiftUI

enum DestinationVideoUiState {
    case idle
    case loading
    case success(destinations: [DestinationWithVideoStatus])
    case error(message: String)
}

enum VideoGenerationState {
    case idle
    case generating
    case success(message: String)
    case error(message: String)
}

@MainActor
class DestinationVideoViewModel: ObservableObject {
    @Published var uiState: DestinationVideoUiState = .idle
    @Published var generationState: VideoGenerationState = .idle
    
    private let service: JourneyService
    
    init(service: JourneyService = JourneyService.shared) {
        self.service = service
    }
    
    func loadUserDestinations(userId: String) async {
        uiState = .loading
        do {
            let response = try await service.getUserDestinations(userId: userId)
            uiState = .success(destinations: response.destinations)
        } catch {
            uiState = .error(message: error.localizedDescription)
            print("❌ [DestinationVideoViewModel] Error loading destinations: \(error.localizedDescription)")
        }
    }
    
    func generateVideo(userId: String, destination: String) async {
        generationState = .generating
        do {
            let response = try await service.generateDestinationVideo(userId: userId, destination: destination)
            generationState = .success(message: response.message)
            // Reload destinations to update status
            await loadUserDestinations(userId: userId)
        } catch {
            generationState = .error(message: error.localizedDescription)
            print("❌ [DestinationVideoViewModel] Error generating video: \(error.localizedDescription)")
        }
    }
    
    func checkVideoStatus(userId: String, destination: String) async {
        do {
            let status = try await service.getDestinationVideoStatus(userId: userId, destination: destination)
            // Update the destination status in the list if needed
            if case .success(var destinations) = uiState {
                if let index = destinations.firstIndex(where: { $0.destination == destination }) {
                    var updatedDestination = destinations[index]
                    // Note: DestinationWithVideoStatus is a struct, so we need to create a new one
                    let newDestination = DestinationWithVideoStatus(
                        destination: updatedDestination.destination,
                        videoStatus: status.status,
                        videoUrl: status.videoUrl,
                        imageCount: status.imageCount,
                        errorMessage: status.errorMessage
                    )
                    destinations[index] = newDestination
                    uiState = .success(destinations: destinations)
                }
            }
        } catch {
            print("❌ [DestinationVideoViewModel] Error checking video status: \(error.localizedDescription)")
        }
    }
}

